flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlToOperationConverterTest.java

```diff
@@ -1466,6 +1477,404 @@ public void testAlterTableCompactOnManagedPartitionedTable() throws Exception {
+                                + "A PRIMARY KEY constraint must be declared on physical columns.");
+
+        ObjectIdentifier tableIdentifier = ObjectIdentifier.of("cat1", "db1", "tb1");
+        Schema originalSchema =
+                catalogManager.getTable(tableIdentifier).get().getTable().getUnresolvedSchema();
+        Operation operation =
+                parse(
+                        "alter table tb1 add constraint my_pk primary key (a, b) not enforced",
+                        SqlDialect.DEFAULT);
+
+        assertAlterTableSchema(
+                operation,
+                tableIdentifier,
+                Schema.newBuilder()
+                        .fromSchema(originalSchema)
+                        .primaryKeyNamed("my_pk", "a", "b")
+                        .build());
+
+        operation =
+                parse(
+                        "alter table tb1 add d bigint not null primary key not enforced",
+                        SqlDialect.DEFAULT);
+        assertAlterTableSchema(
+                operation,
+                tableIdentifier,
+                Schema.newBuilder()
+                        .fromSchema(originalSchema)
+                        .column("d", DataTypes.BIGINT().notNull())
+                        .primaryKey("d")
+                        .build());
+    }
+
+    @Test
+    public void testAlterTableSchemaAddWatermark() throws Exception {
+        prepareNonManagedTable(false);
+
+        // test add watermark with an undefined column as rowtime
+        assertThatThrownBy(
+                        () ->
+                                parse(
+                                        "alter table tb1 add watermark for ts as ts",
+                                        SqlDialect.DEFAULT))
+                .isInstanceOf(ValidationException.class)
+                .hasMessageContaining(
+                        "The rowtime attribute field 'ts' is not defined in the table schema, at line 1, column 35\n"
+                                + "Available fields: ['a', 'b', 'c']");
+
+        // test add watermark with an undefined nested column as rowtime
+        assertThatThrownBy(
+                        () ->
+                                parse(
+                                        "alter table tb1 add (d row<d1 string, ts timestamp(3)>, watermark for d.ts as d.ts)",
+                                        SqlDialect.DEFAULT))
+                .isInstanceOf(ValidationException.class)
+                .hasMessageContaining(
+                        "The nested rowtime attribute field 'd.ts' cannot define a watermark.");
+
+        ObjectIdentifier tableIdentifier = ObjectIdentifier.of("cat1", "db1", "tb1");
+        Schema originalSchema =
+                catalogManager.getTable(tableIdentifier).get().getTable().getUnresolvedSchema();
+
+        // test add watermark with new added physical column as rowtime
```

Comment:
Test add watermark on existed column?

Reply:
> Test add watermark on existed column?

This is covered by `TableEnvironmentTest#testAlterTableAddWatermark`